cpp/include/rmm/detail/cccl_adapters.hpp

```diff
@@ -0,0 +1,269 @@
+
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/detail/cuda_memory_resource.hpp>
+#include <rmm/detail/export.hpp>
+
+namespace RMM_NAMESPACE {
+
+namespace detail {
+
+template <typename ResourceType>
+class cccl_resource_ref : public ResourceType {
+ public:
+  using base = ResourceType;
+
+  using base::base;
+
+  cccl_resource_ref(base const& other) : base(other) {}
+
+  cccl_resource_ref(base&& other) : base(std::move(other)) {}
+
+#if CCCL_MAJOR_VERSION > 3 || (CCCL_MAJOR_VERSION == 3 && CCCL_MINOR_VERSION >= 1)
+
+  void* allocate(std::size_t bytes) { return this->allocate_sync(bytes); }
```

Comment:
I found that this large duplicate code block is much more difficult to track than the previous one. We only need to use SFINAE for the unique functions of each version.

Reply:
Good idea, I agree. I made some tweaks in 2497848e79eac9b941a339c6a1db1ca5602e5c8a and would like to merge this in the current (updated) state.